src/ParamPassing.cc

```diff
@@ -0,0 +1,675 @@
+{
+  ElementPtr elemDesc = initElementDescription(_childrenXml, _errors);
+  if (!elemDesc)
+    return;
+
+  // loop through children and handle corresponding actions
+  for (tinyxml2::XMLElement *xmlChild = _childrenXml->FirstChildElement();
+       xmlChild;
+       xmlChild = xmlChild->NextSiblingElement())
+  {
+    if (xmlChild->Attribute("action") == nullptr)
+    {
+      _errors.push_back({ErrorCode::ATTRIBUTE_MISSING,
+        "Missing an action attribute. Skipping child element modification "
+        "with parent <" + std::string(_childrenXml->Name()) + " element_id='"
+        + std::string(_childrenXml->Attribute("element_id")) + "'>:\n"
+        + ElementToString(xmlChild)
+      });
+      continue;
+    }
+
+    std::string actionStr = xmlChild->Attribute("action");
+    if (!isValidAction(actionStr))
+    {
+      _errors.push_back({ErrorCode::ATTRIBUTE_INVALID,
+        "Action [" + actionStr + "] is not a valid action. Skipping "
+        "child element modification with parent <"
+        + std::string(_childrenXml->Name()) + " element_id='"
+        + std::string(_childrenXml->Attribute("element_id")) + "'>:\n"
+        + ElementToString(xmlChild)
+      });
+      continue;
+    }
+
+    if (actionStr == "remove")
+    {
+      ElementPtr e = getElementByName(_elem, xmlChild);
+      if (e == nullptr)
+      {
+        _errors.push_back({ErrorCode::ELEMENT_MISSING,
+          "Could not find element. Skipping child element removal "
+          "with parent <" + std::string(_childrenXml->Name()) + " element_id='"
+          + std::string(_childrenXml->Attribute("element_id")) + "'>:\n"
+          + ElementToString(xmlChild)
+        });
+      }
+      else
+      {
+        remove(xmlChild, e, _errors);
+      }
+
+      continue;
+    }
+    else if (actionStr == "modify")
+    {
+      ElementPtr e = getElementByName(_elem, xmlChild, true);
+      if (e == nullptr)
+      {
+        _errors.push_back({ErrorCode::ELEMENT_MISSING,
+          "Could not find element. Skipping child element modification "
+          "with parent <" + std::string(_childrenXml->Name()) + " element_id='"
+          + std::string(_childrenXml->Attribute("element_id")) + "'>:\n"
+          + ElementToString(xmlChild)
+        });
+      }
+      else
+      {
+        modify(xmlChild, e, _errors);
+      }
+
+      continue;
+    }
+
+    // get child element description (used for actions: add and replace)
+    bool useParentDesc = false;
+    std::string elemName = xmlChild->Name();
+
+    // the modification will be skipped if not an existing sdf element
+    // TODO(jenn) not sure if we should load element anyway,
+    // might need to be a future implementation
+    if (!elemDesc->HasElementDescription(elemName))
+    {
+      // TODO(anyone) when forward porting to sdf11,
+      // need to add `|| elemName == "model"` since nested models are allowed
+      if (elemName == "link")
+      {
+        useParentDesc = true;
```

Comment:
I'm not quite sure if I understand why `link` is being treated differently here. Is the parent element assumed to be `model`? If so should we include `joint` and `frame` as well?.

Reply:
per VC, we don't need this `if` block. Everything in the `else` block can be hoisted up.

Reply:
3b28f1faece56bd1533ff32a4cb4b85536dded65